Add UB document describing the flat-flow migration (#15597)
dotnet/arcade: Documentation/UnifiedBuild/Flat flow migration guide.md

# Flat dependency flow migration FAQ

> [!NOTE]
> This is part of the [Unified Build](https://github.com/dotnet/arcade/tree/main/Documentation/UnifiedBuild) effort.

This document describes the migration of [VMR repositories](https://github.com/dotnet/dotnet/tree/main/src) from their current Maestro dependency flow to the new VMR-based flat flow.

> This only applies to repositories that are part of the VMR (specifically on `main`).

Reviewer: If I'm repo A that is not part of the VMR but is depending on repo B that is part of the VMR this migration also affects me as my subscriptions will change, no?

Author: Ah yeah, thanks for noticing. This document was named something else originally and now it does not make sense to have this here.

## FAQ

### How can I see dependency subscriptions for my repository?
Either use the [Maestro website](https://maestro.dot.net/subscriptions) or use the [`darc get-subscriptions`](../Darc.md) command.

### My repository is not part of the VMR. Am I still affected?
If your repository has subscriptions to a repository that is [part of the VMR](https://github.com/dotnet/dotnet/tree/main/src), this repository will get built as part of the [VMR build](https://dev.azure.com/dnceng/internal/_build?definitionId=1330).
Your repository will be subscribed to the [VMR](https://github.com/dotnet/dotnet) (`dotnet/dotnet`) instead.
Some VMR repositories will keep their original official build though.

### Where can I find the official build of the VMR?
The official build of the VMR is [dotnet-unified-build](https://dev.azure.com/dnceng/internal/_build?definitionId=1330).

### Where can I find the new dependency PRs?
The dependency PRs look almost the same as the old ones and are still [authored by the `dotnet-maestro` bot](https://github.com/pulls?q=sort%3Aupdated-desc+is%3Apr+author%3Aapp%2Fdotnet-maestro+archived%3Afalse+).  

> The forward flow PRs will be opened [against the VMR](https://github.com/dotnet/dotnet/pulls/app%2Fdotnet-maestro) while the backflow PRs will be opened in your repository and named something like `[branch] Source code changes from dotnet/dotnet`.

Reviewer: Even though forward / backflow PRs are mentioned in the expanded UB terminology, they are not mentioned directly in the Terminology section for this page. Maybe its worth either linking to their definition here or adding them to the Terminology section?

Author: Good points

### Can my VMR repository still subscribe to other repositories?
Yes, you can still set up regular subscriptions to other repositories, **but not to other VMR repositories** - those will flow into your repository through the VMR.

### Can my product repository still produce packages?
Yes, but they should not overlap with those produced in VMR's official build.
[Reach out to us](#whom-to-contact-and-when) if you need help with this.

### What happens to Arcade?

> Arcade's official build will stay as-is and people can depend on it the same way as before.

Reviewer: > people can depend on the same way as before

So this means that its up to the repo to choose how it wants to get Arcade, either directly or from the VMR? If yes, what are the differences between the two? I think this might lead to questions from repos on the best setup for Arcade

> - **every repo will flow** its sources using the new source-enable Maestro subscriptions directly **into the VMR** (under `src/[reponame]`),
> - the **official builds of the product repos will stop producing packages**,
> - **the VMR will become the official build** of the product repos,
> - former dependents on product repositories will depend on and get their packages from the VMR instead (including the product repositories themselves).

The new dependency tree will look like this:

```mermaid
graph TD
    VMR[dotnet/dotnet] --> aspnetcore
    VMR --> other
    VMR --> runtime
    VMR --> sdk

    aspnetcore
    other[...]
    runtime
    sdk

    aspnetcore --> VMR
    other --> VMR
    runtime --> VMR
    sdk --> VMR
```

An in-depth description of this can be found [here](https://github.com/dotnet/arcade/blob/main/Documentation/UnifiedBuild/VMR-Code-And-Build-Workflow.md).

## Migration timeline

- **April 9th** - Migration is announced and introduced to developers in a *.NET Platform Talks* session.
- **April 24th - 26th** - Migration is executed by the Unified Build team.
- **after** - Repository owners are asked to wind down their official builds.

## Migration process

> On the day of migration, the Unified Build team will run a script that will redirect the subscriptions connected to VMR repositories:

Reviewer 2: I am surprised this doesn't mention anything about forward flows. E.g. all forward flows to repos that are included in the VMR get converted into a single flow to the VMR.  All forward flows to non-vmr repos are untouched.

Author: Fair point, let's be more explicit.

    net10 --> wpf[dotnet/wpf]

    winforms -.-> winformsCi(dotnet-winforms-ci<br/>pipeline):::pipeline
    winformsCi -.-> net10(.NET 10 channel):::channel
    net10 --> winformsTests[winforms-tests]

    classDef active fill:#ffcc00,stroke:#aa8800,color:#000;
    classDef pipeline fill:#00ffcc,color:#000;
    classDef channel fill:#00ccff,color:#000;
```

After the migration, the setup will look like this:

```mermaid
graph TD
    winformsAssets[winforms-assets]--> winforms:::active
    
    arcade[dotnet/arcade] --> vmr[dotnet/dotnet]
    runtime[dotnet/runtime] --> vmr
    winforms --> vmr

    vmr -.-> vmrCi(dotnet-unified-build<br/>pipeline):::pipeline
    vmrCi -.-> net10ub(.NET 10 UB channel):::channel

    net10ub --> wpf[dotnet/wpf]
    net10ub --> winformsTests[winforms-tests]
    net10ub --> winforms

    classDef active fill:#ffcc00,stroke:#aa8800,color:#000;
    classDef pipeline fill:#00ffcc,color:#000;
    classDef channel fill:#00ccff,color:#000;
```

The `runtime` and `arcade` dependencies will flow into `winforms` through the VMR, and the `winforms-assets` dependency will still flow directly into `winforms` (as it's not part of the VMR).

> ## FAQ

Reviewer 2: Is it useful to call attention to all backflows happen in one PR?  There is no batching option anymore, it is automatic. The example is to basic to make this obvious.

If your repository has subscriptions to a repository that is [part of the VMR](https://github.com/dotnet/dotnet/tree/main/src), this repository will get built as part of the [VMR build](https://dev.azure.com/dnceng/internal/_build?definitionId=1330).
Your repository will be subscribed to the [VMR](https://github.com/dotnet/dotnet) (`dotnet/dotnet`) instead.
Some VMR repositories will keep their original official build though.

### Where can I find the official build of the VMR?
The official build of the VMR is [dotnet-unified-build](https://dev.azure.com/dnceng/internal/_build?definitionId=1330).

### Where can I find the new dependency PRs?
The dependency PRs look almost the same as the old ones and are still [authored by the `dotnet-maestro` bot](https://github.com/pulls?q=sort%3Aupdated-desc+is%3Apr+author%3Aapp%2Fdotnet-maestro+archived%3Afalse+).  
The forward flow PRs will be opened [against the VMR](https://github.com/dotnet/dotnet/pulls/app%2Fdotnet-maestro) while the backflow PRs will be opened in your repository and named something like `[branch] Source code changes from dotnet/dotnet`.

### Can my VMR repository still subscribe to other repositories?

> Yes, you can still set up regular subscriptions to other repositories, **but not to other VMR repositories** - those will flow into your repository through the VMR.

Reviewer 2: > but not to other VMR repositories

Will our tooling block this?

Author: We will not block this. We will have a mix of VMR repos building in VMR and retaining their official builds so this must not be forbidden.
But I was more verbose and explained this.

### Can my product repository still produce packages?
Yes, but they should not overlap with those produced in VMR's official build.
[Reach out to us](#whom-to-contact-and-when) if you need help with this.

> ### What happens to Arcade?

Reviewer 2: I feel like this section would benefit by having more details.  What is the recommendation here on which to use, what are the pros and cons.  For example for tooling repo's like roslyn that insert into multiple SDK versions, how would they be configured?

Arcade's official build will stay as-is and people can depend on it the same way as before.
It will, however, be also built in VMR's official build so it's possible to subscribe to the VMR too.  
Arcade will also keep its validation loop through `dotnet/arcade-validation`.
It technically does not matter if you depend on `dotnet/dotnet` or `dotnet/arcade` as the packages produced should be practically the same.

### My repo X depends on repo Y's packages. How will I get the new packages?
If repo Y is part of the VMR, you will depend on the VMR instead of repo Y.
The packages will be produced by the official VMR build and published to the `.NET 10 UB` channel.  
The target feed for some packages might change from `dotnet-eng` to [`dotnet10-transport`](https://pkgs.dev.azure.com/dnceng/internal/_packaging/dotnet10-transport/nuget/v3/index.json).

### I have an ongoing dependency PR in my repository with additional changes/work. What should I do?
Finish the PR as you would normally. If there are conflicts with the newly merged backflow PRs, feel free to tag **@dotnet/product-construction** on the PR and we will help you resolve those.

> ### Whom to contact and when?

Reviewer 2: Is there anyplace repo devs can log issues?

# Flat dependency flow migration FAQ

> [!NOTE]
> This is part of the [Unified Build](https://github.com/dotnet/arcade/tree/main/Documentation/UnifiedBuild) effort.

This document describes the migration of [.NET repositories](https://github.com/dotnet/dotnet/tree/main/src) from their current Maestro dependency flow to the new VMR-based flat flow which **will happen on the week of April 24th**.

## Table of contents
- [Terminology](#terminology)
- [High-level overview of the change](#high-level-overview-of-the-change)

> - [Migration timeline](#migration-timeline)

Reviewer 3: Nit: broken link

- **Dependency flow** / **Binary flow** - Old type of Maestro subscriptions that are only flowing dependency updates, e.g. [this one](https://github.com/dotnet/sdk/pull/47085).
- **Code flow** / **Source-enabled dependency flow** - New type of Maestro subscriptions that are, together with dependency updates, also flowing sources to/from the VMR.
- **Backflow** - A PR that flows changes from the VMR back to the product repository, carrying source updates as well as dependency updates (packages built in the VMR).
- **Forward flow** - A PR that flows changes from the product repository to the VMR, carrying repository source updates.
- **Flat flow** - A new structure of subscriptions between product repositories and the VMR.

See also [Unified Build Terminology](./Terminology.md).

## High-level overview of the change

Currently, the .NET SDK is built out of a set of individual repositories which flow dependencies among each other using Maestro dependency update PRs.
The flow ends in `dotnet/sdk` where we can gather the whole dependency tree and assemble the full .NET SDK.  

> [!NOTE]

> > Since .NET 8, we are also mirroring sources representing this dependency tree into the [VMR](https://github.com/dotnet/dotnet) for every commit made to `dotnet/sdk`.

Reviewer 3: NIt: this isn't strictly true. There are plenty of repos in the tree that we don't mirror. E.g. dotnet-wpf-int, or runtime tests, etc.

The owners of the VMR repositories will be asked to wind down their official builds.
This can happen on their own schedule, but it is recommended to do it as soon as possible.

Further documentation will be supplied to repository owners on how to do this but, in essence,
a no-op build for every commit will still be required at first so that the build is still registered in Maestro and published into a channel.
This is so that subscriptions still trigger the same way and we don't alter the infrastructure too much at once.
However, it is expected that in future, no builds will be required and Maestro will trigger subscriptions right after commits mirror to the internal AzDO repositories.

## Example

Let's consider the following made up example where we are migrating the `winforms` repository, which:
- Is part of the VMR,
- Builds officially via the `dotnet-winforms-ci` internal pipeline and is published to the `.NET 10` channel,
- Depends on `arcade` and `runtime` (VMR repositories),
- Depends on `winforms-assets` (a made-up, non-VMR repository),
- Flows into one VMR repository `wpf` and one non-VMR repository `winforms-tests`.

So the setup looks like this:

```mermaid
graph TD

>     arcade[dotnet/arcade] --> winforms:::active

Reviewer 3: I would include channel names on the graph edges

    runtime[dotnet/runtime] --> winforms
    winformsAssets[winforms-assets]--> winforms
    net10 --> wpf[dotnet/wpf]

    winforms -.-> winformsCi(dotnet-winforms-ci<br/>pipeline):::pipeline
    winformsCi -.-> net10(.NET 10 channel):::channel
    net10 --> winformsTests[winforms-tests]

    classDef active stroke:#aa8800;
    classDef pipeline fill:#00ffcc,color:#000;
    classDef channel fill:#00ccff,color:#000;
```

After the migration, the setup will look like this:

```mermaid
graph TD
    winformsAssets[winforms-assets]--> winforms:::active
    
    arcade[dotnet/arcade] --> vmr[dotnet/dotnet]
    runtime[dotnet/runtime] --> vmr
    winforms --> vmr

    vmr -.-> vmrCi(dotnet-unified-build<br/>pipeline):::pipeline
    vmrCi -.-> net10ub(.NET 10 UB channel):::channel

>     net10ub --> wpf[dotnet/wpf]

Reviewer 3: To make a bit more readable, I would have WPF flowing into the VMR

    net10ub --> winformsTests[winforms-tests]
    net10ub --> winforms

    classDef active stroke:#aa8800;
    classDef pipeline fill:#00ffcc,color:#000;
    classDef channel fill:#00ccff,color:#000;
```

The `runtime` and `arcade` dependencies will flow into `winforms` through the VMR, and the `winforms-assets` dependency will still flow directly into `winforms` (as it's not part of the VMR).

## FAQ

### How can I see dependency subscriptions for my repository?
Either use the [Maestro website](https://maestro.dot.net/subscriptions) or use the [`darc get-subscriptions`](../Darc.md) command.

### My repository is not part of the VMR. Am I still affected?

> If your repository has subscriptions to a repository that is [part of the VMR](https://github.com/dotnet/dotnet/tree/main/src), this repository will get built as part of the [VMR build](https://dev.azure.com/dnceng/internal/_build?definitionId=1330).

Reviewer 3: ```suggestion
If your repository has subscriptions to a repository that is [part of the VMR] (https://github.com/dotnet/dotnet/tree/main/src) and is **not** keeping its official build, your repository will be subscribed to the [VMR](https://github.com/dotnet/dotnet) (`dotnet/dotnet`) instead
```

### Where can I find the new dependency PRs?
The dependency PRs look almost the same as the old ones and are still [authored by the `dotnet-maestro` bot](https://github.com/pulls?q=sort%3Aupdated-desc+is%3Apr+author%3Aapp%2Fdotnet-maestro+archived%3Afalse+).  
The forward flow PRs will be opened [against the VMR](https://github.com/dotnet/dotnet/pulls/app%2Fdotnet-maestro) while the backflow PRs will be opened in your repository and named something like `[branch] Source code changes from dotnet/dotnet`.

### Can my VMR repository still subscribe to other repositories?
Yes, you can still set up regular subscriptions to other repositories.
However, **consider subscribing to other VMR repositories** - most of those will turn off their official builds and their packages will flow into your repository from the VMR.
Subscriptions to repositories that will move their official builds into the VMR will be migrated for you during this process.
Subscriptions to repositories that will deprecate their official builds later will change the subscriptions to their dependents then.

### Can my product repository still produce packages?

> Yes, but they should not overlap with those produced in VMR's official build.

Reviewer 3: ```suggestion
Yes, though keep in mind two things:
- These package identities should not overlap with the VMR produced packages. The VMR already takes steps to ensure that version numbers do not overlap.
- Repositories that ship via the dotnet release process (e.g. runtime or aspnetcore) must ship all of their outputs via the VMR.
```

> [!NOTE]
> Since .NET 8, we are also mirroring sources representing this dependency tree into the [VMR](https://github.com/dotnet/dotnet) for every commit made to `dotnet/sdk`.
> Every product repository has [a corresponding folder](https://github.com/dotnet/dotnet/tree/main/src) in the VMR.
The process is described in more detail [here](https://github.com/dotnet/arcade/blob/main/Documentation/UnifiedBuild/VMR-Code-And-Build-Workflow.md#internal-code-flow-and-releases) and [here](https://github.com/dotnet/arcade/blob/main/Documentation/UnifiedBuild/VMR-Design-And-Operation.md#source-synchronization-process).

Today's dependency tree for *product repositories* (minus `arcade`) looks something like this:

```mermaid
graph TD
    sdk --> dotnet
    aspire --> sdk
    aspnetcore --> sdk
    cecil --> runtime
    command-line-api --> runtime
    command-line-api --> sourcelink
    deployment-tools --> sdk
    diagnostics --> vstest
    efcore --> aspnetcore
    emsdk --> runtime
    fsharp --> sdk
    msbuild --> sdk
    nuget-client --> sdk
    razor --> sdk
    roslyn --> sdk
    roslyn-analyzers --> sdk
    runtime --> aspnetcore
    runtime --> efcore

>     runtime --> winforms

Reviewer 4: runtime also flows directly into sdk, not sure if that ruins the graph look tho

Author: Yeah, possibly not important to be 100% precise here. It's rather for illustration

- If your repo depends on several VMR repositories, it will be subscribed to the VMR just once.
  - E.g. if you depend on `runtime` and `aspnetcore`, you will be subscribed to `dotnet/dotnet` only.
    All packages will flow from the VMR together in a single backflow PR.
- VMR will be subscribed to all VMR repositories and their sources will start flowing to the VMR (this is the forward flow).

### Phase II - Official build deprecation

After the main migration is done, VMR repositories will still be building in their internal pipelines.
However, for most of them, this will be extra and unnecessary.
The owners of the VMR repositories will be asked to wind down their official builds.
This can happen on their own schedule, but it is recommended to do it as soon as possible.

Further documentation will be supplied to repository owners on how to do this but, in essence,
a no-op build for every commit will still be required at first so that the build is still registered in Maestro and published into a channel.
This is so that subscriptions still trigger the same way and we don't alter the infrastructure too much at once.

> However, it is expected that in future, no builds will be required and Maestro will trigger subscriptions right after commits mirror to the internal AzDO repositories.

Reviewer 4: ```suggestion
However, it is expected that in the future, no builds will be required and Maestro will trigger subscriptions right after commits mirror to the internal AzDO repositories.
```

    classDef active fill:#ffcc00,stroke:#aa8800,color:#000;
    classDef pipeline fill:#00ffcc,color:#000;
    classDef channel fill:#00ccff,color:#000;
```

After the migration, the setup will look like this:

```mermaid
graph TD
    winformsAssets[winforms-assets]--> winforms:::active

    arcade[dotnet/arcade] --> vmr[dotnet/dotnet]
    runtime[dotnet/runtime] --> vmr
    vmr --> runtime
    vmr --> arcade
    winforms --> vmr

    vmr --> wpf[dotnet/wpf]
    wpf --> vmr
    vmr --> winformsTests[winforms-tests]
    vmr --> winforms

    classDef active fill:#ffcc00,stroke:#aa8800,color:#000;
```

The `runtime` and `arcade` dependencies will flow into `winforms` through the VMR, and the `winforms-assets` dependency will still flow directly into `winforms` (as it's not part of the VMR). The `winforms-tests` repository will newly depend on the VMR instead of `winforms` and will get its packages from the VMR.

## FAQ

### How can I see dependency subscriptions for my repository?
Either use the [Maestro website](https://maestro.dot.net/subscriptions) or use the [`darc get-subscriptions`](../Darc.md) command.

### My repository is not part of the VMR. Am I still affected?

> If your repository has subscriptions to a repository that is [part of the VMR] (https://github.com/dotnet/dotnet/tree/main/src) and is **not** keeping its official build, your repository will be subscribed to the [VMR](https://github.com/dotnet/dotnet) (`dotnet/dotnet`) instead

Reviewer 4: ```suggestion
If your repository has subscriptions to a repository that is [part of the VMR](https://github.com/dotnet/dotnet/tree/main/src) and is **not** keeping its official build, your repository will be subscribed to the [VMR](https://github.com/dotnet/dotnet) (`dotnet/dotnet`) instead
```
